pkg/statedb2/db.go

```diff
@@ -0,0 +1,177 @@
+//  6. If old object existed, remove from revision index
+//  7. If old object existed, remove from graveyard
+//  8. Update each secondary index
+//  9. Commit transaction by committing each index to
+//     the table and then committing table to the root.
+//     Swap the root atomic pointer to new root and
+//     notify by closing channels of all modified nodes.
+//
+// To observe deletions:
+//  1. Create write transaction against the target table
+//  2. Create new delete tracker and add it to the table
+//  3. Commit the write transaction to update the table
+//     with the new delete tracker
+//  4. Query the graveyard by revision, starting from the
+//     revision of the write transaction at which it was
+//     created.
+//  5. For each successfully processed deletion, mark the
+//     revision to set low watermark for garbage collection.
+//  6. Periodically garbage collect the graveyard by finding
+//     the lowest revision of all delete trackers.
+type DB struct {
+	tables    map[TableName]TableMeta
+	mu        lock.Mutex // sequences modifications to the root tree
+	root      atomic.Pointer[iradix.Tree[tableEntry]]
+	gcTrigger chan struct{} // trigger for graveyard garbage collection
+	gcExited  chan struct{}
+}
+
+func NewDB(tables []TableMeta) (*DB, error) {
+	txn := iradix.New[tableEntry]().Txn()
+	db := &DB{
+		tables:    make(map[TableName]TableMeta),
+		gcTrigger: make(chan struct{}, 1),
+		gcExited:  make(chan struct{}),
+	}
+	for _, t := range tables {
+		name := t.Name()
+		if _, ok := db.tables[name]; ok {
+			return nil, fmt.Errorf("table %q already exists", name)
+		}
+		db.tables[name] = t
+		var table tableEntry
+		table.meta = t
+		table.deleteTrackers = iradix.New[deleteTracker]()
+		indexTxn := iradix.New[indexTree]().Txn()
+		indexTxn.Insert([]byte(t.primaryIndexer().name), iradix.New[object]())
+		indexTxn.Insert([]byte(RevisionIndex), iradix.New[object]())
+		indexTxn.Insert([]byte(GraveyardIndex), iradix.New[object]())
+		indexTxn.Insert([]byte(GraveyardRevisionIndex), iradix.New[object]())
+		for index := range t.secondaryIndexers() {
+			indexTxn.Insert([]byte(index), iradix.New[object]())
+		}
+		table.indexes = indexTxn.CommitOnly()
+		txn.Insert(t.tableKey(), table)
+	}
+	db.root.Store(txn.CommitOnly())
+
+	return db, nil
+}
+
+// ReadTxn constructs a new read transaction for performing reads against
+// a snapshot of the database.
+//
+// ReadTxn is not thread-safe!
+func (db *DB) ReadTxn() ReadTxn {
+	return &txn{
+		db:          db,
+		rootReadTxn: db.root.Load().Txn(),
+	}
+}
+
+// WriteTxn constructs a new write transaction against the given set of tables.
+// Each table is locked, which may block until the table locks are acquired.
+// The modifications performed in the write transaction are not visible outside
+// it until Commit() is called. To discard the changes call Abort().
+//
+// WriteTxn is not thread-safe!
+func (db *DB) WriteTxn(table TableMeta, tables ...TableMeta) WriteTxn {
```

Comment:
I think we'll need to be pretty careful about deadlocks when acquiring write transactions. Clearly, the total lock order of the tables prevents deadlocks between tables, but it's still possible for table locks to be included in a wait cycle. Here's an example:
Let's call the ipcache mutex `Mipcache`, and some table Ta's lock `MTa`. If any code path acquires `Mipcache` and then `MTa`, while any other path acquires `Mipcache` while holding `MTa`, we have a deadlock.
It may be prudent to prohibit acquiring _any_ external locks when acquiring a `WriteTxn`. (If not enforceable via some custom linter, we should at least document it).

Reply:
Yep that's a good idea to document! For somewhat simple cases that don't involve function pointers or interfaces we can probably catch this with a linter.